pkg/build/prometheus/metrics.go

```diff
@@ -0,0 +1,172 @@
+	ch <- descActiveBuildCount
+	ch <- descBuildDuration
+	ch <- descStageDuration
+	ch <- descStepDuration
+}
+
+// Collect implements the prometheus.Collector interface.
+func (bc *buildCollector) Collect(ch chan<- prometheus.Metric) {
+	result, err := bc.lister.List(kselector.Everything())
+
+	if err != nil {
+		glog.V(4).Infof("Collect err %v", err)
+		return
+	}
+
+	for _, b := range result {
+		bc.collectBuild(ch, b)
+	}
+
+	glog.V(4).Infof("reconcileActiveBuildMetrics looked at %d builds", len(result))
+
+}
+
+func (bc *buildCollector) collectBuild(ch chan<- prometheus.Metric, b *buildapi.Build) {
+	addGauge := func(desc *prometheus.Desc, v float64, lv ...string) {
+		lv = append([]string{b.ObjectMeta.Namespace, b.ObjectMeta.Name}, lv...)
+		ch <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, v, lv...)
+	}
+	addHistogram := func(desc *prometheus.Desc, v float64, m map[float64]uint64, lv ...string) {
+		lv = append([]string{b.ObjectMeta.Namespace, b.ObjectMeta.Name}, lv...)
+		ch <- prometheus.MustNewConstHistogram(desc, 1, v, m, lv...)
+	}
+
+	creation := b.ObjectMeta.CreationTimestamp.String()
+	phase := strings.ToLower(string(b.Status.Phase))
+	reason := strings.ToLower(string(b.Status.Reason))
+
+	// first, "decrement" this build's active gauges; since we are labeling per build, we can be explicit/precise
+	// and set to zero vs. calling the decrement method
+	addGauge(descActiveBuildCount, 0, new, creation)
+	addGauge(descActiveBuildCount, 0, pending, creation)
+	addGauge(descActiveBuildCount, 0, running, creation)
+
+	// now, based on the actual phase for each build, update the various metrics for that single (i.e. 1) build; again, since we are labeling
+	// per build, we can be explicit/precise vs. calling the increment method
+	switch b.Status.Phase {
+	case buildapi.BuildPhaseNew:
+		addGauge(descActiveBuildCount, 1, new, creation)
+	case buildapi.BuildPhasePending:
+		addGauge(descActiveBuildCount, 1, pending, creation)
+	case buildapi.BuildPhaseRunning:
+		addGauge(descActiveBuildCount, 1, running, creation)
+	case buildapi.BuildPhaseFailed:
+		addGauge(descTerminalBuildCount, 1, phase, reason)
+	case buildapi.BuildPhaseError:
+		addGauge(descTerminalBuildCount, 1, phase, reason)
+	case buildapi.BuildPhaseCancelled:
+		addGauge(descTerminalBuildCount, 1, phase, reason)
+	case buildapi.BuildPhaseComplete:
+		addGauge(descTerminalBuildCount, 1, phase, reason)
+		for _, stage := range b.Status.Stages {
+			if stage.DurationMilliseconds != 0 {
+				durationInSec := float64(stage.DurationMilliseconds) / float64(1000)
+				stagebucket := map[float64]uint64{float64(1): uint64(durationInSec)}
+				addHistogram(descStageDuration, durationInSec, stagebucket, strings.ToLower(string(stage.Name)))
+			}
+			for _, step := range stage.Steps {
+				if step.DurationMilliseconds != 0 {
+					durationInSec := float64(stage.DurationMilliseconds) / float64(1000)
+					stepbucket := map[float64]uint64{float64(1): uint64(durationInSec)}
```

Comment:
Allocating a map per step is way too much memory.  You should allocate only one map per collection, not one per step per stage per object